In [5]:
##Set Path
path = r'/Users/heatherstewart/Dropbox/Mac/Desktop/Instacart Basket analysis 3'

In [6]:
path

'/Users/heatherstewart/Dropbox/Mac/Desktop/Instacart Basket analysis 3'

In [7]:
##import orders csv file
df_ords = pd.read_csv(os.path.join(path, '02 Data','Original Data','orders.csv'),index_col = False)

In [8]:
##import product csv file
df_prods = pd.read_csv(os.path.join(path,'02 Data','Original Data','products.csv'),index_col = False)

In [9]:
##Getting a sense of df_ords dataframe by looking at the head
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [10]:
##Getting a sense of df_prods dataframe by looking at the head
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [12]:
##Dropping columns in df_ords dataframe that will not be useful to my analysis
df_ords.drop(columns =['eval_set','order_id'])

user_id  order_number  order_dow  order_hour_of_day  \
0              1             1          2                  8   
1              1             2          3                  7   
2              1             3          3                 12   
3              1             4          4                  7   
4              1             5          4                 15   
...          ...           ...        ...                ...   
3421078   206209            10          5                 18   
3421079   206209            11          4                 10   
3421080   206209            12          1                 12   
3421081   206209            13          1                 12   
3421082   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 5 columns]

In [13]:
##Assigning the new dataframe without the columns. 
df_ords = df_ords.drop(columns = ['eval_set','order_id'])

In [14]:
df_ords.head()

user_id  order_number  order_dow  order_hour_of_day  days_since_prior_order
0        1             1          2                  8                     NaN
1        1             2          3                  7                    15.0
2        1             3          3                 12                    21.0
3        1             4          4                  7                    29.0
4        1             5          4                 15                    28.0

In [15]:
##Step 3: Find an unintuitive name and change its name without overwriting the dataframe.
df_ords.rename (columns ={'order_dow':'order_day_of_week', 'order_number':'order_category_number'},inplace = True)

In [16]:
df_ords.head()

user_id  order_category_number  order_day_of_week  order_hour_of_day  \
0        1                      1                  2                  8   
1        1                      2                  3                  7   
2        1                      3                  3                 12   
3        1                      4                  4                  7   
4        1                      5                  4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [17]:
##Step 4: Tell the client the busiest hour for placing order by finding the frequency of the corresponding variable
df_ords['order_hour_of_day'].value_counts()

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

10 and 11am are the busiest hours of the day, followed by 3,2,1pm and 12 noon. 

In [18]:
##Step 5: Determine the meaning of behind the value of 4 in department_id within the df_prods dataframe.
df_dept = pd.read_csv(os.path.join(path, '02 Data','Original Data','departments.csv'),index_col = False)

In [19]:
df_dept.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

##Transposing dataframe-turning dataframe rows into columns(changing dataframe from wide format into long format.

In [20]:
df_dept.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [21]:
##Assigning the dataframe to long format
df_dept = df_dept.T

In [22]:
##Adding index to department dataframe
df_dept.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [23]:
##Creating new header for department dataframe
new_header = df_dept.iloc[0]

In [24]:
new_header

0    department
Name: department_id, dtype: object

In [25]:
##Deleting first row to prevent double headers
df_dept_new = df_dept[1:]

In [26]:
df_dept_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [29]:
##Using ".columns" to set new headers.
df_dept_new.columns = new_header 

In [30]:
df_dept_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

## DATA DICTIONARY

In [31]:
data_dict = df_dept_new.to_dict('index')

In [32]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [33]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [34]:
print(data_dict.get('4'))

{'department': 'produce'}


##The number "4" is the produce department. 

## CREATING SUBSET

In [35]:
##Step 6: The sales team wants to know more about breakfast items.Create a subset containing the information.
df_breakfast = df_prods[df_prods['department_id']==14]

In [36]:
df_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

In [37]:
##Step 7: ##Step 7: Details on products for dinner such as alcohol, deli, beverages and meat/seafood
df_dinner = df_prods.loc[df_prods['department_id'].isin ([5, 12, 20,7])]

In [38]:
df_dinner

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

## Keeping track of the dataframe

In [39]:
##Step 8: How many rows does the last dataframe have?

In [40]:
df_dinner.shape

(7650, 5)

The last dataframe (df_dinner) has 7650 rows and 5 columns. 

In [41]:
##Step 9: Extracting all the information on customer with the "user_id" "1".

In [42]:
df_ords_user1 = df_ords[df_ords['user_id']==1]

In [43]:
df_ords_user1

user_id  order_category_number  order_day_of_week  order_hour_of_day  \
0         1                      1                  2                  8   
1         1                      2                  3                  7   
2         1                      3                  3                 12   
3         1                      4                  4                  7   
4         1                      5                  4                 15   
5         1                      6                  2                  7   
6         1                      7                  1                  9   
7         1                      8                  1                 14   
8         1                      9                  1                 16   
9         1                     10                  4                  8   
10        1                     11                  4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

In [44]:
df_ords_user1.info

<bound method DataFrame.info of     user_id  order_category_number  order_day_of_week  order_hour_of_day  \
0         1                      1                  2                  8   
1         1                      2                  3                  7   
2         1                      3                  3                 12   
3         1                      4                  4                  7   
4         1                      5                  4                 15   
5         1                      6                  2                  7   
6         1                      7                  1                  9   
7         1                      8                  1                 14   
8         1                      9                  1                 16   
9         1                     10                  4                  8   
10        1                     11                  4                  8   

    days_since_prior_order  
0                      NaN

In [45]:
df_ords_user1.describe()

user_id  order_category_number  order_day_of_week  order_hour_of_day  \
count     11.0              11.000000          11.000000          11.000000   
mean       1.0               6.000000           2.636364          10.090909   
std        0.0               3.316625           1.286291           3.477198   
min        1.0               1.000000           1.000000           7.000000   
25%        1.0               3.500000           1.500000           7.500000   
50%        1.0               6.000000           3.000000           8.000000   
75%        1.0               8.500000           4.000000          13.000000   
max        1.0              11.000000           4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

In [46]:
df_ords_user1.columns

Index(['user_id', 'order_category_number', 'order_day_of_week',
       'order_hour_of_day', 'days_since_prior_order'],
      dtype='object')

In [47]:
##Step 10:What basic statistical information can be provided based on the previous information
## Customer with user_id 1 has 11 days recorded for the week--this need to be investigated. 

## EXPORTING DATAFRAME

In [ ]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))
